In [2]:
!java --version
!python --version

openjdk 11.0.18 2023-01-17
OpenJDK Runtime Environment (build 11.0.18+10-post-Ubuntu-0ubuntu120.04.1)
OpenJDK 64-Bit Server VM (build 11.0.18+10-post-Ubuntu-0ubuntu120.04.1, mixed mode, sharing)
Python 3.9.16


In [3]:
!pip install pyspark

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 281.4/281.4 MB 4.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 199.7/199.7 kB 22.2 MB/s eta 0:00:00
  Created wheel for pyspark: filename=pyspark-3.3.2-py2.py3-none-any.whl size=281824028 sha256=b5578f2a5827e69b201c1ffe0ae9129d367db07fdb93a72cfcf5d6f5e7cb25dc
  Stored in directory: /root/.cache/pip/wheels/6c/e3/9b/0525ce8a69478916513509d43693511463c6468db0de237c86
Successfully built pyspark
  Attempting uninstall: py4j
    Found existing installation: py4j 0.10.9.7
    Uninstalling py4j-0.10.9.7:
      Successfully uninstalled py4j-0.10.9.7


In [4]:
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.recommendation import ALS
from pyspark.sql import Row

In [7]:
from pyspark.context import SparkContext
from pyspark.sql.session import SparkSession
sc = SparkContext('local')
spark = SparkSession(sc)

In [9]:
lines = spark.read.text("sample_movielens_ratings.txt").rdd
parts = lines.map(lambda row: row.value.split("::"))
ratingsRDD = parts.map(lambda p: Row(userId=int(p[0]), movieId=int(p[1]),
                                     rating=float(p[2]), timestamp=int(p[3])))
ratings = spark.createDataFrame(ratingsRDD)
(training, test) = ratings.randomSplit([0.8, 0.2])

In [18]:
# Build the recommendation model using ALS on the training data
# Note we set cold start strategy to 'drop' to ensure we don't get NaN evaluation metrics
als = ALS(maxIter=5, regParam=0.01, userCol="userId", itemCol="movieId", ratingCol="rating",
          coldStartStrategy="drop")
model = als.fit(training)

In [19]:
# Evaluate the model by computing the RMSE on the test data
predictions = model.transform(test)
evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating",
                                predictionCol="prediction")
rmse = evaluator.evaluate(predictions)
print("Root-mean-square error = " + str(rmse))

Root-mean-square error = 1.8118317090496183


In [23]:
# Build the recommendation model using ALS on the training data
# Note we set cold start strategy to 'drop' to ensure we don't get NaN evaluation metrics
als2 = ALS(maxIter=5, regParam=0.1, userCol="userId", itemCol="movieId", ratingCol="rating",
          coldStartStrategy="drop")
model2 = als2.fit(training)

In [24]:
# Evaluate the model by computing the RMSE on the test data
predictions2 = model2.transform(test)
evaluator2 = RegressionEvaluator(metricName="rmse", labelCol="rating",
                                predictionCol="prediction")
rmse = evaluator2.evaluate(predictions2)
print("Root-mean-square error of maxIter 5 and regParam 0.1 = " + str(rmse))

Root-mean-square error of maxIter 5 and regParam 0.1 = 1.128928739579148


In [21]:
# Build the recommendation model using ALS on the training data
# Note we set cold start strategy to 'drop' to ensure we don't get NaN evaluation metrics
als3 = ALS(maxIter=10, regParam=0.5, userCol="userId", itemCol="movieId", ratingCol="rating",
          coldStartStrategy="drop")
model3 = als3.fit(training)

In [22]:
predictions3 = model3.transform(test)
evaluator3 = RegressionEvaluator(metricName="rmse", labelCol="rating",
                                predictionCol="prediction")
rmse = evaluator3.evaluate(predictions3)
print("Root-mean-square error of maxIter 10 and regParam 0.5 = " + str(rmse))

Root-mean-square error of maxIter 10 and regParam 0.5 = 1.3661437934262444


In [32]:
# Build the recommendation model using ALS on the training data
# Note we set cold start strategy to 'drop' to ensure we don't get NaN evaluation metrics
als4 = ALS(maxIter=10, regParam=1, userCol="userId", itemCol="movieId", ratingCol="rating",
          coldStartStrategy="drop")
model4 = als4.fit(training)

In [38]:
predictions4 = model4.transform(test)
evaluator4 = RegressionEvaluator(metricName="rmse", labelCol="rating",
                                predictionCol="prediction")
rmse = evaluator4.evaluate(predictions4)
print("Root-mean-square error of maxIter 10 and regParam 1.0 = " + str(rmse))

Root-mean-square error of maxIter 10 and regParam 1.0 = 1.648753463817064


In [34]:
# Build the recommendation model using ALS on the training data
# Note we set cold start strategy to 'drop' to ensure we don't get NaN evaluation metrics
als5 = ALS(maxIter=10, regParam=0.1, userCol="userId", itemCol="movieId", ratingCol="rating",
          coldStartStrategy="drop")
model5 = als5.fit(training)

In [35]:
predictions5 = model5.transform(test)
evaluator5 = RegressionEvaluator(metricName="rmse", labelCol="rating",
                                predictionCol="prediction")
rmse = evaluator5.evaluate(predictions5)
print("Root-mean-square error of maxIter 20 and regParam 0.1 = " + str(rmse))

Root-mean-square error of maxIter 20 and regParam 0.1 = 1.1024369296176486


In [36]:
# Build the recommendation model using ALS on the training data
# Note we set cold start strategy to 'drop' to ensure we don't get NaN evaluation metrics
als6 = ALS(maxIter=10, regParam=0.01, userCol="userId", itemCol="movieId", ratingCol="rating",
          coldStartStrategy="drop")
model6 = als6.fit(training)

In [37]:
predictions6 = model6.transform(test)
evaluator6 = RegressionEvaluator(metricName="rmse", labelCol="rating",
                                predictionCol="prediction")
rmse = evaluator6.evaluate(predictions6)
print("Root-mean-square error of maxIter 20 and regParam 0.1 = " + str(rmse))

Root-mean-square error of maxIter 20 and regParam 0.1 = 1.6727653465889263


In [39]:
# Build the recommendation model using ALS on the training data
# Note we set cold start strategy to 'drop' to ensure we don't get NaN evaluation metrics
als7 = ALS(maxIter=5, regParam=0.5, userCol="userId", itemCol="movieId", ratingCol="rating",
          coldStartStrategy="drop")
model7 = als7.fit(training)

In [40]:
predictions7 = model7.transform(test)
evaluator7 = RegressionEvaluator(metricName="rmse", labelCol="rating",
                                predictionCol="prediction")
rmse = evaluator7.evaluate(predictions7)
print("Root-mean-square error of maxIter 5 and regParam 0.5 = " + str(rmse))

Root-mean-square error of maxIter 5 and regParam 0.5 = 1.3654758054384069


In [41]:
# Build the recommendation model using ALS on the training data
# Note we set cold start strategy to 'drop' to ensure we don't get NaN evaluation metrics
als8 = ALS(maxIter=5, regParam=1.0, userCol="userId", itemCol="movieId", ratingCol="rating",
          coldStartStrategy="drop")
model8 = als8.fit(training)

In [42]:
predictions8 = model8.transform(test)
evaluator8 = RegressionEvaluator(metricName="rmse", labelCol="rating",
                                predictionCol="prediction")
rmse = evaluator8.evaluate(predictions8)
print("Root-mean-square error of maxIter 5 and regParam 1.0 = " + str(rmse))

Root-mean-square error of maxIter 5 and regParam 1.0 = 1.6487507218372317


Conclusion:

maxIter 5:

regParam 0.01 = 1.8118317090496183 (Model 1)

reParam 0.1 = 1.128928739579148 (Model 2)

reParam 0.5 = 1.3654758054384069 (Model 7)

reParam 1 = 1.6487507218372317 (Model 8)




maxIter 10:

regParam 1 = 1.648753463817064 (Model 3)

reParam 0.5 = 1.3661437934262444 (Model 4)

reParam 0.1 = 1.1024369296176486 (Model 5)

reParam 0.01 = 1.6727653465889263 (Model 6)



From Models above we can see that the models perform better depends on maxIter and regParam where regParam closer to 0.1 and the larger the maxIter ( 5 < 10 ) produce better RMSE value 

In [20]:
# Generate top 10 movie recommendations for each user
userRecs = model.recommendForAllUsers(10)

In [13]:
# Generate top 10 user recommendations for each movie
movieRecs = model.recommendForAllItems(10)

In [14]:
# Generate top 10 movie recommendations for a specified set of users
users = ratings.select(als.getUserCol()).distinct().limit(3)
userSubsetRecs = model.recommendForUserSubset(users, 10)

In [15]:
# Generate top 10 user recommendations for a specified set of movies
movies = ratings.select(als.getItemCol()).distinct().limit(3)
movieSubSetRecs = model.recommendForItemSubset(movies, 10)